src/infrastructure/api/index.ts

```diff
@@ -0,0 +1,112 @@
+import AWS from 'aws-sdk';
+import express, { Request } from 'express';
+import mysql from 'mysql2/promise';
+import vehicleQueryFunctionFactory from '../../app/vehicleQueryFunctionFactory';
+import testQueryFunctionFactory from '../../app/testQueryFunctionFactory';
+import { getResultsDetails, getVehicleDetails } from '../../domain/enquiryService';
+import ParametersError from '../../errors/ParametersError';
+import ResultsEvent from '../../interfaces/ResultsEvent';
+import VehicleEvent from '../../interfaces/VehicleEvent';
+import DatabaseService from '../databaseService';
+import SecretsManagerService from '../secretsManagerService';
+
+const app = express();
+const router = express.Router();
+
+const { API_VERSION } = process.env;
+
+// Declare middlewares
+/**
+ * bodyParser, error handling, logger, etc..
+ * http://expressjs.com/en/starter/basic-routing.html
+ * http://expressjs.com/en/guide/using-middleware.html
+ */
+
+/**
+ * app level middlewares
+ * app.use('/path', (req, res, next) => {
+ * chain middlewares
+ * next()
+ * })
+ */
+router.use((req, _response, next) => {
+  // TODO Add logger lib like Winston or Morgan
+  console.log('path');
+  console.log(req.path);
+  next();
+});
+
+// Debug router before we start proxying  requests from /v<x> psth
+router.get('/', (_request, res) => {
+  res.send({ ok: true });
+});
+
+router.get('/version', (_request, res) => {
+  res.send({ version: API_VERSION });
+});
+
+router.get(
+  '/vehicle',
+  (
+    request: Request<Record<string, unknown>, string | Record<string, unknown>, Record<string, unknown>, VehicleEvent>,
+    res,
+  ) => {
+    console.info('Handling vehicle request');
+    let secretsManager;
+    try {
+      secretsManager = new SecretsManagerService(new AWS.SecretsManager());
+    } catch (e) {
+      if (e instanceof Error) {
+        res.status(500).send(e.message);
+      }
+    }
+
+    DatabaseService.build(secretsManager, mysql)
```

Comment:
Can we try to use async await instead so let's say in the future if we have to perform more async tasks in  sequence we avoid nesting promises, etc..

Reply:
Unfortunately if you try using async/await in these functions typescript complains because there isn't a type definition for them that returns a promise, so we're forced to use normal promise handling.

Reply:
Had a go and managed to make it work however I am not sure why the linter is complaining when the unit tests run.. some errors have been turned into warning to bypass that. It is quite odd as this is 'legit' express syntax when manipulating the `req` & `res` objects..

```diff
diff --git a/.eslintrc.js b/.eslintrc.js

index cb2b270..73bd674 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -11,5 +11,14 @@ module.exports = {
     'jest/no-identical-title': 'error',
     'jest/prefer-to-have-length': 'warn',
     'jest/valid-expect': 'error',
+    '@typescript-eslint/no-misused-promises': [
+      'error',
+      {
+        checksVoidReturn: false,
+      },
+    ],
+    '@typescript-eslint/no-unsafe-member-access': 'warn',
+    '@typescript-eslint/no-unsafe-call': 'warn',
+    '@typescript-eslint/no-unsafe-assignment': 'warn',
   },
 };
diff --git a/sonar-project.properties b/sonar-project.properties
index ca71996..90f9012 100644
--- a/sonar-project.properties
+++ b/sonar-project.properties
@@ -6,7 +6,7 @@
 # Encoding of the source code. Default is default system encoding
 sonar.sourceEncoding=UTF-8

-# sonar.host.url=http://volted-monkey-sonarqube.ops:9000
+# sonar.host.url=http://localhost:9000
 sonar.host.url=http://volted-monkey-sonarqube.ops:9000

 # must be unique in a given SonarQube instance
@@ -24,3 +24,5 @@ sonar.exclusions=node_modules/**, /**/*.test.ts, coverage/**, local/**, .vscode/
 sonar.test.inclusions=*.test.ts
 sonar.javascript.lcov.reportPaths=coverage/unit/lcov.info
 sonar.eslint.reportPaths=reports/eslint/eslint-report.json
+sonar.login=
+sonar.password=
diff --git a/src/infrastructure/api/index.ts b/src/infrastructure/api/index.ts
index 9fdc828..4b7a426 100644
--- a/src/infrastructure/api/index.ts
+++ b/src/infrastructure/api/index.ts
@@ -1,14 +1,15 @@
 import AWS from 'aws-sdk';
 import express, { Request } from 'express';
 import mysql from 'mysql2/promise';
+
 import vehicleQueryFunctionFactory from '../../app/vehicleQueryFunctionFactory';
 import testQueryFunctionFactory from '../../app/testQueryFunctionFactory';
-import { getResultsDetails, getVehicleDetails } from '../../domain/enquiryService';
+import { getResultsDetails } from '../../domain/enquiryService';
 import ParametersError from '../../errors/ParametersError';
 import ResultsEvent from '../../interfaces/ResultsEvent';
-import VehicleEvent from '../../interfaces/VehicleEvent';
 import DatabaseService from '../databaseService';
 import SecretsManagerService from '../secretsManagerService';
+import * as Controllers from '../../interfaces/controllers';

 const app = express();

@@ -44,38 +45,7 @@ app.get('/version', (_request, res) => {
   res.send({ version: API_VERSION });
 });

-app.get(
-  '/enquiry/vehicle',
-  (
-    request: Request<Record<string, unknown>, string | Record<string, unknown>, Record<string, unknown>, VehicleEvent>,
-    res,
-  ) => {
-    console.info('Handling vehicle request');
-    let secretsManager;
-    try {
-      secretsManager = new SecretsManagerService(new AWS.SecretsManager());
-    } catch (e) {
-      if (e instanceof Error) {
-        res.status(500).send(e.message);
-      }
-    }
-
-    DatabaseService.build(secretsManager, mysql)
-      .then((dbService) => getVehicleDetails(request.query, vehicleQueryFunctionFactory, dbService))
-      .then((result) => {
-        res.contentType('json').send(result);
-      })
-      .catch((e: Error) => {
-        if (e instanceof ParametersError) {
-          res.status(400);
-        } else {
-          res.status(500);
-        }
-
-        res.send(e.message);
-      });
-  },
-);
+app.get('/enquiry/vehicle', Controllers.vehicleController(mysql, vehicleQueryFunctionFactory, DatabaseService));

 app.get(
   '/enquiry/testResults',
diff --git a/src/interfaces/controllers/index..ts b/src/interfaces/controllers/index..ts
deleted file mode 100644
index e69de29..0000000
```
also extracted some logic of the controller into a separate module and used dependency injection for the DB, etc.. in case later we wanted to remove it so it is no longer part of the app

```js
// interfaces/controllers/index.ts

export * from './VehicleController';
```

```js
// interfaces/controllers/VehicleController.ts

import AWS from 'aws-sdk';
import { Request } from 'express';

import { getVehicleDetails } from '../../domain/enquiryService';
import ParametersError from '../../errors/ParametersError';
import SecretsManagerService from '../../infrastructure/secretsManagerService';
import VehicleEvent from '../VehicleEvent';

export const vehicleController = (mysql, vehicleQueryFunctionFactory, databaseService) => async (
  request: Request<Record<string, unknown>, string | Record<string, unknown>, Record<string, unknown>, VehicleEvent>,
  res,
) => {
  console.info('Handling vehicle request');
  let secretsManager;
  try {
    secretsManager = new SecretsManagerService(new AWS.SecretsManager());
    const dbService = await databaseService.build(secretsManager, mysql);
    const result = await getVehicleDetails(request.query, vehicleQueryFunctionFactory, dbService);
    res.contentType('json').send(result);
  } catch (e) {
    if (e instanceof ParametersError) {
      res.status(400);
    } else {
      res.status(500);
    }
    res.send(e.message);
  }
};

```

What do you think?

Reply:
It looks fine to me, although I find the controller syntax a bit weird to follow. I was trying to do it without changing the linting rules as well because I'm only here for a short amount of time and didn't want to change anything that might affect you guys in the future. 

However given that what is there works and the release decision is today, I'd rather put it this improvement into a backlog ticket and fix it in the future, once we've got a release out.

Reply:
yes, the linter is complaining about the mutation of the request, res object which is pretty standard when using express so we may need to tweak them.... It probably makes sense to create a ticket in the future in case the routes and controllers grow we may want to separate them if it is not for now.